From dcd1d9ab1462b49093465e3e4117419b63778799 Mon Sep 17 00:00:00 2001
From: =?UTF-8?q?R=C3=A9mi=20Bernon?= <rbernon@codeweavers.com>
Date: Mon, 22 Mar 2021 18:11:27 +0100
Subject: [PATCH] user32: Send WM_INPUT_DEVICE_CHANGE / RAWINPUT to the server.

Expect INPUT/INPUT_HARDWARE to be followed with RAWINPUT, when uMsg is
WM_INPUT_DEVICE_CHANGE. This is for internal __wine_send_input calls
only, as we ignore INPUT_HARDWARE in SendInput.

Wine-Bug: https://bugs.winehq.org/show_bug.cgi?id=50506
---
 dlls/user32/message.c | 19 +++++++++++++++++++
 1 file changed, 19 insertions(+)

diff --git a/dlls/user32/message.c b/dlls/user32/message.c
index f87ef9fb3af..f11aaa4144d 100644
--- a/dlls/user32/message.c
+++ b/dlls/user32/message.c
@@ -3267,6 +3267,25 @@ NTSTATUS send_hardware_message( HWND hwnd, const INPUT *input, const RAWINPUT *r
         case INPUT_HARDWARE:
             req->input.hw.msg    = input->u.hi.uMsg;
             req->input.hw.lparam = MAKELONG( input->u.hi.wParamL, input->u.hi.wParamH );
+            switch (input->u.hi.uMsg)
+            {
+            case WM_INPUT_DEVICE_CHANGE:
+                req->input.hw.data.rawinput.type = rawinput->header.dwType;
+                switch (rawinput->header.dwType)
+                {
+                case RIM_TYPEHID:
+                    assert( rawinput->data.hid.dwCount <= 1 );
+                    req->input.hw.data.rawinput.hid.device = HandleToUlong( rawinput->header.hDevice );
+                    req->input.hw.data.rawinput.hid.param = rawinput->header.wParam;
+                    req->input.hw.data.rawinput.hid.length = rawinput->data.hid.dwSizeHid;
+                    if (rawinput->data.hid.dwSizeHid)
+                        wine_server_add_data( req, rawinput->data.hid.bRawData, rawinput->data.hid.dwSizeHid );
+                    break;
+                default:
+                    assert( 0 );
+                    break;
+                }
+            }
             break;
         }
         if (key_state_info) wine_server_set_reply( req, key_state_info->state,
-- 
2.30.2

